examples/manifold_gradients.dx

```diff
@@ -0,0 +1,342 @@
+
+' We now define the exponential and logarithm map for the quaternions. The
+exponential map at the identity is the well-known mapping from axis-angle to
+quaternions.  Away from the identity, we can use the fact that SO3 is a Lie
+group to transport the exponential map at the identity to the appropriate
+tangent space, by using simple right-multiplication by $q$ to associate the
+two tangent spaces.
+
+' Since we need to ensure that the exponential function is differentiable at
+zero (and similar for the logarithm), we can't use standard formulations of
+functions such as `sinc x = (sin x) / x` or
+`magnitude x = sqrt $ sum (for i. x.i * x.i)`. To get around this, we use some
+simple Taylor series expansions, noting that this will cause reduced precision,
+and running autodiff on the resulting linearizations may be inefficient.
+
+-- Convenience alias for an array of three floats.
+R3 = (Fin 3 => Float)
+
+-- Taylor expansion for sinc (sqrt x).
+def sincSqrt (x : Float) : Float = 
+ (1.
+  - (x  / 6.)
+  + (x * x / 120.)
+  - (x * x * x / 5040.)
+  + (x * x * x * x / 362880.)
+  - (x * x * x * x * x/ 39916800.)
+  + (x * x * x * x * x *  x / 6227020800.))
+
+-- Taylor expansion for cos (sqrt x).
+def cosSqrt (x: Float) : Float = 
+  (1. 
+   - (x / 2.)
+   + (x * x / 24.)
+   - (x * x * x / 720.)
+   + (x * x * x * x / 40320.) 
+   - (x * x * x * x * x / 3628800.)
+   + (x * x * x * x * x * x / 479001600.))
+
+-- Taylor expansion for sinc (arccos x)
+def sincAcos (x: Float) : Float = 
+  (2. / pi
+   + x * 4. / (pow pi 2.)
+   + x * x * (8. - pow pi 2.) / (pow pi 3.)
+   - x * x * x * 4. * (pow pi 2. - 12.) / (3. * pow pi 4.)
+   + x * x * x * x * (384. - 16. * pow pi 2. - 3. * pow pi 4.) / (12. * pow pi 5.)
+   - x * x * x * x * x * 8. * (pow pi 4. - 120.) / (15. * pow pi 6.))
+
+-- Exponential function at the identity.
+def quatExpAtIdent (v : R3) : Quaternion = 
+  [x', y', z'] = v
+  m2 = squareMagnitude v
+  -- This is equal to sinc (m / 2) = 2 * sin (m / 2) / m
+  sincHalfM = sincSqrt $ m2 / 4.
+  Q {x = x' * sincHalfM / 2.,
+     y = y' * sincHalfM / 2.,
+     z = z' * sincHalfM / 2.,
+     w = cosSqrt $ m2 / 4. }
+
+-- Inverse to exponential function at the identity.
+def quatExpAtIdentInv (q: Quaternion) : R3 =
+  -- Fine to use qNormalize here, since q should be approximately unit magnitude
+  -- (or at least non-zero), so the sqrt operation can be differentiated.
+  (Q{x, y, z, w}) = quatNormalize q
+  2. .* [x, y, z] / (sincAcos w)
+
+
+instance TangentSpace Quaternion R3
+  tangentExponentialMap =
+    \ q v . quatExpAtIdent v * q
+  tangentLogarithmMap =
+    \ q p . quatExpAtIdentInv (p * quatInv q)
+
+
+' Run some simple checks to make sure that the exponential and logarithm maps
+have the correct properties.
+
+-- Just arbitrary quaternions.
+quatForTestOne = quatNormalize (Q{w=1.0, x=0.5, y=-0.2, z=0.4})
+quatForTestTwo = quatNormalize (Q{w=-0.4, x=0.5, y=0.8, z=-0.9})
+
+-- Should be approximately equal to quatForTestOne.
```

Comment:
I'd recommend using `~~` here as well.

Reply:
The tolerance is required to be a bit lower for numerical reasons, but I've changed to using `allclose`.